app/api-client.js

```diff
@@ -1,170 +1,291 @@
-const config = require('../app/config');
+const config = require('./config');
 const request = require('request-promise');
 const url = require('url');
 
 
-module.exports = (function () {
-  var token = 'invalid token';
+const log = require('bole')('api-client');
+
+
+class JWTAuth {
+  constructor(token) {
+    this.token = token || 'invalid token';
+  }
+
+  set_token(token) {
+    this.token = token;
+  }
+
+  unset_token() {
+    this.token = 'invalid token';
+  }
+
+  get header() {
+    return `JWT ${this.token}`;
+  }
+}
+
+
+class APIClient {
+  constructor(conf) {
+    this.base_url = conf.base_url;
+    this.auth = null;
+  }
+
+  request(endpoint, method = 'GET', body = null) {
+    const headers = {};
+
+    if (this.auth) {
+      headers.Authorization = this.auth.header;
+    }
 
-  function api_request(options) {
     try {
-      return request(override_defaults(options));
+      const options = {
+        method,
+        uri: this.endpoint_url(endpoint),
+        headers,
+        body,
+        json: true,
+      };
+      return request(options)
+        .then((result) => {
```

Comment:
So we're resolving the request promise here. Does it mean this `request()` method will be blocking?

I still need to check the rest of the file...sorry if it's a silly question.

Reply:
`.then()` actually returns another Promise, that's why you can chain `.then()` and `.catch()`.